lib/index.js

```diff
@@ -0,0 +1,146 @@
+'use strict'
+const Promise = require('bluebird')
+const yaml = require('js-yaml')
+const Joi = require('joi')
+const hostname = require('@runnable/hostname')
+
+const removeNullAndUndefinedKeys = require('./util').removeNullAndUndefinedKeys
+const validate = require('./util').validate
+const Parser = require('./parser')
+const Warning = require('./warning')
+
+const dockerComposeFileSchema = Joi.object({
+  version: Joi.string().valid('2').required(),
+  services: Joi.object().required()
+}).unknown().required()
+
+const arrayOfStrings = Joi.array().items(Joi.string())
+
+const serviceSchemaKeys = {
+  build: Joi.alternatives().try(Joi.string(), Joi.object()),
+  image: Joi.string(), // TODO: Add parser
+  command: Joi.alternatives().try(Joi.string(), arrayOfStrings),
+  ports: arrayOfStrings,
+  expose: arrayOfStrings, // TODO
+  links: arrayOfStrings, // TODO:
+  depends_on: arrayOfStrings, // TODO
+  environment: arrayOfStrings
+}
+const serviceSchema = Joi.object(serviceSchemaKeys).unknown().required()
+
+module.exports = class DockerComposeParser {
+
+  static parse (opts) {
+    const dockerComposeFileString = opts.dockerComposeFileString
+    const repositoryName = opts.repositoryName
+    const userContentDomain = opts.userContentDomain
+    const ownerUsername = opts.ownerUsername
+    return Promise.try(() => {
+      const dockerCompose = yaml.safeLoad(dockerComposeFileString)
+      return validate(dockerComposeFileSchema, dockerCompose)
+      .then(dockerCompose => dockerCompose.services)
+    })
+    .then(services => {
+      const mainServiceName = DockerComposeParser.getMainName(services)
+      // Turn into an array
+      return Object.keys(services).map(serviceKey => {
+        const links = services[serviceKey].links || []
+        const dependsOn = services[serviceKey].depends_on || []
+        return {
+          metadata: {
+            name: serviceKey,
+            isMain: serviceKey === mainServiceName,
+            links: links.concat(dependsOn)
+          },
+          raw: services[serviceKey]
+        }
+      })
+    })
+    .map(service => {
```

Comment:
NNNNNNNOOOOO!!!!! I actually want to avoid that. I want to make this thing as stupid as possible. It takes a string and it return an objects. That's it. You can pass the object to api client and create an instance/cv. Boom!

This basically makes the module a much simpler transformation function that has no state. I think this will be easier to maintain in the long run, because I'm sure will have some weird edge cases with some docker compose files.